src/memory_tracker.h

```diff
@@ -197,6 +197,15 @@ class MemoryTracker {
   inline void Track(const MemoryRetainer* retainer,
                     const char* edge_name = nullptr);
 
+  // Use when a retainer is embedded in another.
+  // Put a memory container into the graph, create an edge from
+  // the current node if there is one on the stack - there should
+  // be one, of the container object which the current field is part of.
+  // Reduce the size of memory from the container so that retentions
+  // are accounted properly.
+  inline void TrackInlineField(const MemoryRetainer* retainer,
```

Comment:
I am not sure what this is trying to achieve, is there a use case of this API?

Normally the size of the pointer to the retainer should be accounted into the self size of the parent. The self size of the retainer should not be accounted into the parent, so I can't tell when it's necessary to do a subtraction - maybe there is a misuse of the API?

Reply:
If this is trying to address the issues that came up from https://github.com/nodejs/node/pull/26099/ there are two ways to solve that:

1. Do not use `SET_SELF_SIZE(Worker)`, which lazily uses `sizeof(Worker)` to calculate the self size of the worker - that's where we encounter the first tracking of the inline field. Instead, override `Worker::SelfSize()` to exclude the size of the non-pointer fields.
2. Keep using `SET_SELF_SIZE(Worker)` for the initial calculation, but rename this method into something like `SplitInlineField`, because this does not add more sizes into the tracker - instead it's splitting of existing sizes that have already been tracked.

Reply:
@joyeecheung - thanks.

Probably these many scenarios exists (with a component `cmp` and a container `cnt` with `cnt` a retainer):
 - `cmp` is a normal primitive field: self size of `cnt` covers size of `cmp`
 - `cmp` is a composite object (with or without primitive fields): self size of `cnt` covers size of `cmp`
 - `cmp` is a pointer to a composite object: self size of `cnt` covers size of `cmp` (size of a pointer)
 - `cmp` is a composite object (with pointer type fields): self size of `cnt` covers size of `cmp`, not the dynamic memory [ example: messageport in worker ]
 - `cmp` is a composite retainer object (with pointer types): [ current use case ]
 - `cmp` is a pointer to a retainer object (with pointer types): [ none at the moment? ]

We have `Worker` with 200 bytes excluding 2 inline `AsyncRequest` fields of 72 bytes each. So the question is: should we add those to the worker (because those belongs to worker) or add separately - as they are retainers themselves.

If I follow your approach 1, we need to perform that in all such scenarios? in which case arguably the parent retainer (Worker's referencing object if any) could do the same for worker, and potentially render the Retainer abstraction meaningless?

second proposal looks fair to me, but would love to hear from @addaleax as well. Can `SplitStackAllocatedField` be somewhat misleading, as there is no memory that is stack-allocated here?

Reply:
> because this does not necessarily add more memory into the tracker - instead it's splitting off existing memory that have already been tracked, and potentially adding more (if the `retainer` overrides `MemoryInfo()` to add additional fields that are not accounted by a `sizeof()`)

I’m not sure, but I don’t really see this as a issue/as not “tracking” these fields? Then again, re-naming is not a big deal, so I’m okay with anything.

Reply:
> should we add those to the worker (because those belongs to worker) or add separately - as they are retainers themselves.

That depends on whether we want to display AsyncRequest separately as a Node referenced by the Worker node, and any other fields referenced by those AsyncRequests recursively.

> If I follow your approach 1, we need to perform that in all such scenarios? in which case arguably the parent retainer (Worker's referencing object if any) could do the same for worker, and potentially render the Retainer abstraction meaningless?

The abstraction helps adding nodes recursively instead of tracking everything from the root, allowing the MemoryRetainer implementations define how they want their fields to be tracked.

> Can `SplitStackAllocatedField` be somewhat misleading, as there is no memory that is stack-allocated here?

Right, the fields are not really stack allocated if the whole thing is not stack allocated.

> I’m not sure, but I don’t really see this as a issue/as not “tracking” these fields?

I am fine with the current naming if the comments are changed to reflect when this method should be used and what issue it is trying to address. `Use when a retainer is embedded in another.` and `Reduce the size of memory from the container so that retentions are accounted properly.` is too ambiguous IMO, the core of the issue this method is trying to solve arise from that the parent does not implement its `SelfSize()` correctly to exclude the memory that is supposed to be tracked as the self size of another node.

Reply:
@joyeecheung - thanks, I will amend the comments, aligning with above conversation.

Reply:
@joyeecheung - amended the comment, ptal!